api/signup.go

```diff
@@ -16,6 +16,61 @@ type SignupParams struct {
 	Email    string                 `json:"email"`
 	Password string                 `json:"password"`
 	Data     map[string]interface{} `json:"data"`
+	Provider string                 `json:"provider"`
+	Code     string                 `json:"code"`
+}
+
+func (a *API) signupExternalProvider(ctx context.Context, w http.ResponseWriter, r *http.Request, params *SignupParams) {
+	provider, err := a.Provider(params.Provider)
+	if err != nil {
+		BadRequestError(w, fmt.Sprintf("Unsupported provider: %+v", err))
+		return
+	}
+
+	tok, err := provider.GetOAuthToken(ctx, params.Code)
+	if err != nil {
+		InternalServerError(w, fmt.Sprintf("Unable to exchange external code"))
+		return
+	}
+
+	aud := a.requestAud(ctx, r)
+	params.Email, err = provider.GetUserEmail(ctx, tok)
+	if err != nil {
+		InternalServerError(w, fmt.Sprintf("Error getting user email from external provider"))
+		return
+	}
+
+	if exists, err := a.db.IsDuplicatedEmail(params.Email, aud); exists {
+		BadRequestError(w, "User already exists")
+		return
+	} else if err != nil {
+		InternalServerError(w, fmt.Sprintf("Error checking for duplicate users: %+v", err))
+		return
+	}
+
+	user, err := a.signupNewUser(params, aud)
+	if err != nil {
+		InternalServerError(w, err.Error())
+		return
+	}
+
+	if a.config.Mailer.Autoconfirm {
+		user.Confirm()
+	} else if user.ConfirmationSentAt.Add(time.Minute * 15).Before(time.Now()) {
```

Comment:
Yeah exactly, this is just to prevent the possibility of DOSing someone's inbox. Without the timeout it could be possible to fill someone else's inbox with password reset requests.